### PR TITLE
[JENKINS-70988] `TypeError: inputs[i].nextElementSibling is null`

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-70988](https://issues.jenkins.io/browse/JENKINS-70988). Might also correct some problem with the Gradle plugin reported as [JENKINS-70739](https://issues.jenkins.io/browse/JENKINS-70739), though I did not check that. There is a lot of discussion of the need for unique `id` attributes, but I am not actually sure what that is all about. The apparent reason that radio buttons did not work is just that radio groups did not get unique `name` attributes. This is normally ensured by prepending `removeme123-` to the `name`, tracked as `radioPrefix`. That logic was however broken in the case of `credentials/select.jelly` at least by #7631 because it presumed that `nextElementSibling` was defined when this can be null when custom radio button usages are present. When that happened, the browser console would show

```
TypeError: inputs[i].nextElementSibling is null
    <anonymous> http://localhost:8080/jenkins/adjuncts/f24ce642/lib/form/repeatable/repeatable.js:267
    applySubtree http://localhost:8080/jenkins/scripts/behavior.js:116
    applySubtree http://localhost:8080/jenkins/scripts/behavior.js:111
    applySubtree http://localhost:8080/jenkins/scripts/behavior.js:93
    apply http://localhost:8080/jenkins/scripts/behavior.js:76
    start http://localhost:8080/jenkins/scripts/behavior.js:71
    onload http://localhost:8080/jenkins/scripts/behavior.js:133
    onload http://localhost:8080/jenkins/scripts/behavior.js:133
```

The error aborted the loop, causing only the first radio group to get a uniquified radio prefix.

### Testing done

Install `credentials-binding`. Define secret text credentials. Create a freestyle project. Add a credentials binding wrapper and define three bindings of all secret text type (does not seem to matter which). Save and reconfigure. You will see radio buttons misbehaving: the first group (choice of two radio buttons at the top of a binding entry) is fine but the buttons in the second and third binding are improperly mixed up (since all are named `_.credentialsId` with no prefix).

With this patch, each group gets its own `name` as expected, and as happened prior to #7631.

### Proposed changelog entries

- Fix radio buttons in repeated blocks in configuration forms (regression in 2.391).

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7823"></a>

### CODE REVIEW

**Review by MarkEWaite — APPROVED** [review]
I tested credentials binding as described by Jesse and confirmed I can see the problem on the master branch, including the null message in the developer console of the browser.  I confirmed that the fix does not report the null message in the developer console of the browser and that the changes are preserved.

I also tested the text operations plugin in the "before" and "after" cases.  I saw the issue in the "before" case and confirmed that the issue is resolved in the "after" case.

Running the job in the "before" case would generate null pointer exceptions and fail the job if I had previously used the job configuration page and not set all the radio buttons again.  The UI became confused when I tried to adjust one of the radio buttons during those edits.

Running the job in the "after" case did not generate null pointer exceptions whether I had edited the job configuration or not.  The UI did not become confused by adjustments to the radio buttons.

I'd like one more reviewer, but I believe this is ready to merge even without one more reviewer.

**Review by NotMyFault — APPROVED** [review]
Thanks, Jesse.

I could successfully verify this proposal with the gradle plugin affected.

**Comment by MarkEWaite** [comment]
This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.

**Review by uhafner — APPROVED** [review]
Yes, this is almost the same fix I suggested in https://github.com/jenkinsci/jenkins/pull/7734#discussion_r1141300188.

Thanks for fixing this! I can confirm that the selection in the warnings plugin still works.

I would suggest to make the `id` property in the corresponding jelly forms mandatory though. Then we can prevent similar errors in the future.

**Comment by jglick** [comment]
@uhafner https://github.com/jenkinsci/jenkins/pull/7734#discussion_r1141300188 would probably also be prudent; the current code adds a bogus `id` attribute to [the controls](https://github.com/jenkinsci/credentials-plugin/blob/c23caa9a2cb0381fa2c62bb53bed2cb9aad0201c/src/main/resources/lib/credentials/select.jelly#L71-L80) which would otherwise have none, and this is perhaps worse than nothing since both buttons in the block get the same `id` which violates the normal semantics for this attribute. Perhaps that would also avoid the `TypeError` as a side effect, but I preferred to directly address the unsafe assumption, which could also apply to an `input` with an `id` but no `nextElementSibling` if there are any.

> make the `id` property in the corresponding jelly forms mandatory

Which “form” are you referring to? The Jelly fragment noted above creates an `input` element directly, not via any taglib, if that is what you meant. `f:radio` already enforces an `id`: https://github.com/jenkinsci/jenkins/blob/26b8c770e65d79fe2018922a50015774ce774dbb/core/src/main/resources/lib/form/radio.jelly#L46-L50

> the need for unique `id` attributes

My vague understanding is that this was about `label` elements, as in https://github.com/jenkinsci/jenkins/blob/26b8c770e65d79fe2018922a50015774ce774dbb/core/src/main/resources/lib/form/radio.jelly#L52 I think the bug being addressed here applies to hand-written `input`s _not_ using `f:radio`.

By the way, I believe https://github.com/jenkinsci/jenkins/blob/26b8c770e65d79fe2018922a50015774ce774dbb/core/src/main/resources/lib/form/radio.jelly#L32 is incorrect or at least badly misleading. The caller of this tag is _not_ expected to uniquify the `name`; it is normally a simple `fieldName` (or `_.fieldName`), and it is https://github.com/jenkinsci/jenkins/blob/26b8c770e65d79fe2018922a50015774ce774dbb/core/src/main/resources/lib/form/repeatable/repeatable.js#L246-L253 which automatically uniquifies the `name` as the page is loaded. I barely recall d660d60ea7e2e411972b215fc7f6fee9ff03e329; perhaps the `radioPrefix` trick was broken at the time.

**Comment by uhafner** [comment]
I see, the `id` is generated if not set, so my comment is wrong. I must admit that I actually did not really understand what is happening under the hood when these repeatables are created and placed into the page. The `repeatable.js` seems to be quite complex (and still contains YUI code?)...

**Pull request merged**
